[Storage] Uploading list of files to bucket fails

Seems the following test failed, after this #1494, cc @infwinston:
```
_________________________________________________________________________________________________ TestStorageWithCredentials.test_list_source[StoreType.S3] __________________________________________________________________________________________________
[gw0] linux -- Python 3.10.6 /home/ubuntu/.conda/envs/sky/bin/python3.1
tests/test_smoke.py:1361: in test_list_source
    tmp_local_list_storage_obj.add_store(store_type)
sky/data/storage.py:658: in add_store
    self._sync_store(store)
sky/data/storage.py:725: in _sync_store
    os.path.join(self.source, '.git')):
../.conda/envs/sky/lib/python3.10/posixpath.py:76: in join
    a = os.fspath(a)
E   TypeError: expected str, bytes or os.PathLike object, not list
```

I believe the reason is that we did not take the `self.source` to be a list into account in the following place. https://github.com/skypilot-org/skypilot/blob/742c1dc82e62322d4e6d61c68b021b816f6110ad/sky/data/storage.py#L723-L728

I ran into this while running smoke tests. Could we fix it soon @infwinston?
```
FAILED tests/test_smoke.py::TestStorageWithCredentials::test_list_source[StoreType.S3] - TypeError: expected str, bytes or os.PathLike object, not list
FAILED tests/test_smoke.py::TestStorageWithCredentials::test_list_source[StoreType.GCS] - TypeError: expected str, bytes or os.PathLike object, not list
```

> I ran into this while running smoke tests. Could we fix it soon @infwinston?
> 
> ```
> FAILED tests/test_smoke.py::TestStorageWithCredentials::test_list_source[StoreType.S3] - TypeError: expected str, bytes or os.PathLike object, not list
> FAILED tests/test_smoke.py::TestStorageWithCredentials::test_list_source[StoreType.GCS] - TypeError: expected str, bytes or os.PathLike object, not list
> ```

#1519 is fixing this. It will be merged soon. ;)

--- a/.github/workflows/mypy-generic.yml
+++ b/.github/workflows/mypy-generic.yml
@@ -1,0 +1,22 @@
+# This is needed for GitHub Actions for the "Waiting for status to be reported" problem,
+# according to https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/troubleshooting-required-status-checks
+name: mypy
+
+on:
+  # Trigger the workflow on push or pull request,
+  # but only for the main branch
+  push:
+    branches:
+      - master
+      - 'releases/**'
+  pull_request:
+    branches:
+      - master
+      - 'releases/**'
+jobs:
+  mypy:
+    # Need to specify 20.04, because ubuntu-latest does not work with
+    # python 3.6: https://github.com/actions/setup-python/issues/355#issuecomment-1335042510
+    runs-on: ubuntu-20.04
+    steps:
+      - run: 'echo "No mypy to run"'

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -1,4 +1,4 @@
-name: yapf
+name: mypy
 
 on:
   # Trigger the workflow on push or pull request,
@@ -12,7 +12,7 @@ on:
       - master
       - 'releases/**'
 jobs:
-  yapf:
+  mypy:
     # Need to specify 20.04, because ubuntu-latest does not work with
     # python 3.6: https://github.com/actions/setup-python/issues/355#issuecomment-1335042510
     runs-on: ubuntu-20.04
@@ -31,6 +31,6 @@ jobs:
       run: |
         python -m pip install --upgrade pip
         pip install mypy==0.991
-    - name: Running yapf
+    - name: Running mypy
       run: |
         mypy ${{ matrix.mypy-path }}

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -1,0 +1,36 @@
+name: yapf
+
+on:
+  # Trigger the workflow on push or pull request,
+  # but only for the main branch
+  push:
+    branches:
+      - master
+      - 'releases/**'
+  pull_request:
+    branches:
+      - master
+      - 'releases/**'
+jobs:
+  yapf:
+    # Need to specify 20.04, because ubuntu-latest does not work with
+    # python 3.6: https://github.com/actions/setup-python/issues/355#issuecomment-1335042510
+    runs-on: ubuntu-20.04
+    strategy:
+      matrix:
+        python-version: ["3.9"]
+        mypy-path:
+          - sky/data/storage.py
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install mypy==0.991
+    - name: Running yapf
+      run: |
+        mypy ${{ matrix.mypy-path }}

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -19,8 +19,6 @@ jobs:
     strategy:
       matrix:
         python-version: ["3.8"]
-        mypy-path:
-          - sky/data/storage.py
     steps:
     - uses: actions/checkout@v2
     - name: Set up Python ${{ matrix.python-version }}
@@ -30,7 +28,7 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install mypy==0.991
+        pip install mypy==$(grep mypy requirements-dev.txt | cut -d'=' -f3)
     - name: Running mypy
       run: |
-        mypy ${{ matrix.mypy-path }}
+        mypy @tests/mypy_files.txt

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -18,7 +18,7 @@ jobs:
     runs-on: ubuntu-20.04
     strategy:
       matrix:
-        python-version: ["3.9"]
+        python-version: ["3.8"]
         mypy-path:
           - sky/data/storage.py
     steps:

--- a/format.sh
+++ b/format.sh
@@ -34,10 +34,10 @@ tool_version_check() {
     fi
 }
 
-tool_version_check "yapf" $YAPF_VERSION "0.32.0"
-tool_version_check "pylint" $PYLINT_VERSION "2.8.2"
-tool_version_check "pylint-quotes" $PYLINT_QUOTES_VERSION "0.2.3"
-tool_version_check "mypy" "$MYPY_VERSION" "0.991"
+tool_version_check "yapf" $YAPF_VERSION "$(grep yapf requirements-dev.txt | cut -d'=' -f3)"
+tool_version_check "pylint" $PYLINT_VERSION "$(grep "pylint==" requirements-dev.txt | cut -d'=' -f3)"
+tool_version_check "pylint-quotes" $PYLINT_QUOTES_VERSION "$(grep "pylint-quotes==" requirements-dev.txt | cut -d'=' -f3)"
+tool_version_check "mypy" "$MYPY_VERSION" "$(grep mypy requirements-dev.txt | cut -d'=' -f3)"
 
 YAPF_FLAGS=(
     '--recursive'
@@ -47,23 +47,6 @@ YAPF_FLAGS=(
 YAPF_EXCLUDES=(
     '--exclude' 'sky/skylet/providers/**'
 )
-
-# TODO(zhwu): When more of the codebase is typed properly, the mypy flags
-# should be set to do a more stringent check.
-# Keep sync with .github/workflows/mypy.yml
-MYPY_FILES=(
-    # Relative to python/ray
-    'sky/data/storage.py'
-)
-
-# Runs mypy on each argument in sequence. This is different than running mypy
-# once on the list of arguments.
-mypy_on_each() {
-    for file in "$@"; do
-       echo "Running mypy on $file"
-       mypy "$file"
-    done
-}
 
 # Format specified files
 format() {
@@ -93,7 +76,6 @@ format_all() {
     yapf --in-place "${YAPF_FLAGS[@]}" "${YAPF_EXCLUDES[@]}" sky tests examples
 }
 
-
 ## This flag formats individual files. --files *must* be the first command line
 ## arg to use this option.
 if [[ "$1" == '--files' ]]; then
@@ -106,6 +88,12 @@ else
    # Format only the files that changed in last commit.
    format_changed
 fi
+
+# Run mypy
+# TODO(zhwu): When more of the codebase is typed properly, the mypy flags
+# should be set to do a more stringent check.
+echo 'SkyPilot mypy:'
+mypy @tests/mypy_files.txt
 
 # Run Pylint
 echo 'Sky Pylint:'
@@ -121,7 +109,3 @@ if ! git diff --quiet &>/dev/null; then
 
     exit 1
 fi
-
-# Run mypy
-echo 'SkyPilot MYPY:'
-mypy_on_each "${MYPY_FILES[@]}"

--- a/format.sh
+++ b/format.sh
@@ -24,6 +24,7 @@ builtin cd "$ROOT" || exit 1
 YAPF_VERSION=$(yapf --version | awk '{print $2}')
 PYLINT_VERSION=$(pylint --version | head -n 1 | awk '{print $2}')
 PYLINT_QUOTES_VERSION=$(pip list | grep pylint-quotes | awk '{print $2}')
+MYPY_VERSION=$(mypy --version | awk '{print $2}')
 
 # # params: tool name, tool version, required version
 tool_version_check() {
@@ -36,6 +37,7 @@ tool_version_check() {
 tool_version_check "yapf" $YAPF_VERSION "0.32.0"
 tool_version_check "pylint" $PYLINT_VERSION "2.8.2"
 tool_version_check "pylint-quotes" $PYLINT_QUOTES_VERSION "0.2.3"
+tool_version_check "mypy" "$MYPY_VERSION" "0.991"
 
 YAPF_FLAGS=(
     '--recursive'
@@ -45,6 +47,25 @@ YAPF_FLAGS=(
 YAPF_EXCLUDES=(
     '--exclude' 'sky/skylet/providers/**'
 )
+
+# TODO(zhwu): When more of the codebase is typed properly, the mypy flags
+# should be set to do a more stringent check.
+# Keep sync with .github/workflows/mypy.yml
+MYPY_FILES=(
+    # Relative to python/ray
+    'sky/data/storage.py'
+)
+
+# Runs mypy on each argument in sequence. This is different than running mypy
+# once on the list of arguments.
+mypy_on_each() {
+    pushd python/ray
+    for file in "$@"; do
+       echo "Running mypy on $file"
+       mypy "$file"
+    done
+    popd
+}
 
 # Format specified files
 format() {
@@ -74,6 +95,7 @@ format_all() {
     yapf --in-place "${YAPF_FLAGS[@]}" "${YAPF_EXCLUDES[@]}" sky tests examples
 }
 
+
 ## This flag formats individual files. --files *must* be the first command line
 ## arg to use this option.
 if [[ "$1" == '--files' ]]; then
@@ -101,3 +123,7 @@ if ! git diff --quiet &>/dev/null; then
 
     exit 1
 fi
+
+# Run mypy
+echo 'SkyPilot MYPY:'
+mypy_on_each "${MYPY_FILES[@]}"

--- a/format.sh
+++ b/format.sh
@@ -59,12 +59,10 @@ MYPY_FILES=(
 # Runs mypy on each argument in sequence. This is different than running mypy
 # once on the list of arguments.
 mypy_on_each() {
-    pushd python/ray
     for file in "$@"; do
        echo "Running mypy on $file"
        mypy "$file"
     done
-    popd
 }
 
 # Format specified files

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -16,3 +16,7 @@ env = [
     "SKYPILOT_DEBUG=1",
     "SKYPILOT_DISABLE_USAGE_COLLECTION=1"
 ]
+
+[tool.mypy]
+python_version = "3.9"
+

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -19,4 +19,5 @@ env = [
 
 [tool.mypy]
 python_version = "3.8"
-
+follow_imports = "skip"
+ignore_missing_imports = true

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -18,5 +18,5 @@ env = [
 ]
 
 [tool.mypy]
-python_version = "3.9"
+python_version = "3.8"
 

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -5,6 +5,9 @@ pylint==2.8.2
 pylint-quotes==0.2.3
 toml==0.10.2
 
+# type checking
+mypy==0.991
+
 # testing
 pytest
 pytest-xdist

--- a/sky/adaptors/azure.py
+++ b/sky/adaptors/azure.py
@@ -13,7 +13,7 @@ def import_package(func):
         global azure
         if azure is None:
             try:
-                import azure as _azure
+                import azure as _azure # type: ignore
                 azure = _azure
             except ImportError:
                 raise ImportError('Fail to import dependencies for Azure.'

--- a/sky/adaptors/azure.py
+++ b/sky/adaptors/azure.py
@@ -13,7 +13,7 @@ def import_package(func):
         global azure
         if azure is None:
             try:
-                import azure as _azure # type: ignore
+                import azure as _azure  # type: ignore
                 azure = _azure
             except ImportError:
                 raise ImportError('Fail to import dependencies for Azure.'

--- a/sky/data/data_utils.py
+++ b/sky/data/data_utils.py
@@ -110,7 +110,7 @@ def _group_files_by_dir(
     return grouped_files, dirs
 
 
-def parallel_upload(source_path_list: List[Path],
+def parallel_upload(source_path_list: List[str],
                     filesync_command_generator: Callable[[str, List[str]], str],
                     dirsync_command_generator: Callable[[str, str], str],
                     bucket_name: str,

--- a/sky/data/data_utils.py
+++ b/sky/data/data_utils.py
@@ -3,7 +3,6 @@
 from multiprocessing import pool
 import os
 import subprocess
-from pathlib import Path
 from typing import Any, Callable, Dict, List, Optional, Tuple
 import urllib.parse
 

--- a/sky/data/storage.py
+++ b/sky/data/storage.py
@@ -3,7 +3,8 @@ import enum
 import os
 import subprocess
 import time
-from typing import Any, Dict, Optional, Tuple, Union, List
+import typing
+from typing import Any, Dict, Optional, Tuple, Type, Union, List
 import urllib.parse
 
 import colorama
@@ -20,6 +21,10 @@ from sky import exceptions
 from sky import global_user_state
 from sky import sky_logging
 from sky.utils import ux_utils
+
+if typing.TYPE_CHECKING:
+    import boto3  # type: ignore
+    from google.cloud import storage  # type: ignore
 
 logger = sky_logging.init_logger(__name__)
 
@@ -125,7 +130,7 @@ class AbstractStore:
         def __init__(self,
                      *,
                      name: str,
-                     source: str,
+                     source: Union[Path, List[Path], None],
                      region: Optional[str] = None,
                      is_sky_managed: Optional[bool] = None):
             self.name = name
@@ -142,7 +147,7 @@ class AbstractStore:
 
     def __init__(self,
                  name: str,
-                 source: str,
+                 source: Union[Path, List[Path], None],
                  region: Optional[str] = None,
                  is_sky_managed: Optional[bool] = None):
         """Initialize AbstractStore
@@ -221,25 +226,25 @@ class AbstractStore:
         """
         raise NotImplementedError
 
-    def download_remote_dir(self, local_path: str) -> None:
-        """Downloads directory from remote bucket to the specified
-        local_path
+    # def download_remote_dir(self, local_path: str) -> None:
+    #     """Downloads directory from remote bucket to the specified
+    #     local_path
 
-        Args:
-          local_path: Local path on user's device
-        """
-        assert local_path is not None
-        local_path = os.path.expanduser(local_path)
-        iterator = self._remote_filepath_iterator()
-        for remote_path in iterator:
-            remote_path = next(iterator)
-            if remote_path[-1] == '/':
-                continue
-            path = os.path.join(local_path, remote_path)
-            if not os.path.exists(os.path.dirname(path)):
-                os.makedirs(os.path.dirname(path))
-            logger.info(f'Downloading {remote_path} to {path}')
-            self._download_file(remote_path, path)
+    #     Args:
+    #       local_path: Local path on user's device
+    #     """
+    #     assert local_path is not None
+    #     local_path = os.path.expanduser(local_path)
+    #     iterator = self._remote_filepath_iterator()
+    #     for remote_path in iterator:
+    #         remote_path = next(iterator)
+    #         if remote_path[-1] == '/':
+    #             continue
+    #         path = os.path.join(local_path, remote_path)
+    #         if not os.path.exists(os.path.dirname(path)):
+    #             os.makedirs(os.path.dirname(path))
+    #         logger.info(f'Downloading {remote_path} to {path}')
+    #         self._download_file(remote_path, path)
 
     def _download_file(self, remote_path: str, local_path: str) -> None:
         """Downloads file from remote to local on Store
@@ -308,7 +313,7 @@ class Storage(object):
             self,
             *,
             storage_name: Optional[str],
-            source: Optional[str],
+            source: Union[str, List[str], None],
             sky_stores: Optional[Dict[StoreType,
                                       AbstractStore.StoreMetadata]] = None):
             assert storage_name is not None or source is not None
@@ -339,7 +344,7 @@ class Storage(object):
                  stores: Optional[Dict[StoreType, AbstractStore]] = None,
                  persistent: Optional[bool] = True,
                  mode: StorageMode = StorageMode.MOUNT,
-                 sync_on_reconstruction: Optional[bool] = True):
+                 sync_on_reconstruction: bool = True) -> None:
         """Initializes a Storage object.
 
         Three fields are required: the name of the storage, the source
@@ -377,8 +382,6 @@ class Storage(object):
             there. This is set to false when the Storage object is created not
             for direct use, e.g. for sky storage delete.
         """
-        self.name = name
-        self.source = source
         self.persistent = persistent
         self.mode = mode
         assert mode in StorageMode
@@ -390,19 +393,19 @@ class Storage(object):
         self.force_delete = False
 
         # Validate and correct inputs if necessary
-        self._validate_storage_spec()
+        self.name, self.source = self._validate_storage_spec(name, source)
 
         # Sky optimizer either adds a storage object instance or selects
         # from existing ones
         self.stores = {} if stores is None else stores
 
         # Logic to rebuild Storage if it is in global user state
-        self.handle = global_user_state.get_handle_from_storage_name(self.name)
-        if self.handle:
+        handle = global_user_state.get_handle_from_storage_name(self.name)
+        if handle is not None:
             # Reconstruct the Storage object from the global_user_state
             logger.debug('Detected existing storage object, '
                          f'loading Storage: {self.name}')
-            for s_type, s_metadata in self.handle.sky_stores.items():
+            for s_type, s_metadata in handle.sky_stores.items():
                 # When initializing from global_user_state, we override the
                 # source from the YAML
                 if s_type == StoreType.S3:
@@ -433,11 +436,11 @@ class Storage(object):
             sky_managed_stores = {
                 t: s.get_metadata()
                 for t, s in self.stores.items()
-                if s.is_sky_managed()
+                if s.is_sky_managed
             }
-            self.handle = self.StorageMetadata(storage_name=self.name,
-                                               source=self.source,
-                                               sky_stores=sky_managed_stores)
+            handle = self.StorageMetadata(storage_name=self.name,
+                                          source=self.source,
+                                          sky_stores=sky_managed_stores)
 
             if self.source is not None:
                 # If source is a pre-existing bucket, connect to the bucket
@@ -447,10 +450,13 @@ class Storage(object):
                         self.add_store(StoreType.S3)
                     elif self.source.startswith('gs://'):
                         self.add_store(StoreType.GCS)
+        assert handle is not None, 'handle should not be None'
+        self.handle = handle
 
     @staticmethod
-    def _validate_source(source: str, mode: StorageMode,
-                         sync_on_reconstruction: bool) -> Tuple[str, bool]:
+    def _validate_source(
+            source: Union[str, List[str]], mode: StorageMode,
+            sync_on_reconstruction: bool) -> Tuple[Union[str, List[str]], bool]:
         """Validates the source path.
 
         Args:
@@ -545,21 +551,25 @@ class Storage(object):
                         f'Supported paths: local, s3://, gs://. Got: {source}')
         return source, is_local_source
 
-    def _validate_storage_spec(self) -> None:
+    def _validate_storage_spec(
+        self,
+        name: Optional[str],
+        source: Union[Path, List[Path], None],
+    ) -> Tuple[str, Union[Path, List[Path], None]]:
         """
         Validates the storage spec and updates local fields if necessary.
         """
-        if self.source is None:
+        if source is None:
             # If the mode is COPY, the source must be specified
             if self.mode == StorageMode.COPY:
                 # Check if a Storage object already exists in global_user_state
                 # (e.g. used as scratch previously). Such storage objects can be
                 # mounted in copy mode even though they have no source in the
                 # yaml spec (the name is the source).
-                handle = global_user_state.get_handle_from_storage_name(
-                    self.name)
+                handle = global_user_state.get_handle_from_storage_name(name)
                 if handle is not None:
-                    return
+                    assert name is not None
+                    return name, source
                 else:
                     with ux_utils.print_exception_no_traceback():
                         raise exceptions.StorageSourceError(
@@ -569,31 +579,32 @@ class Storage(object):
                 # If source is not specified in COPY mode, the intent is to
                 # create a bucket and use it as scratch disk. Name must be
                 # specified to create bucket.
-                if not self.name:
+                if not name:
                     with ux_utils.print_exception_no_traceback():
                         raise exceptions.StorageSpecError(
                             'Storage source or storage name must be specified.')
                 else:
                     # Create bucket and mount
-                    return
-        elif self.source is not None:
+                    return name, source
+        elif source is not None:
             source, is_local_source = Storage._validate_source(
-                self.source, self.mode, self.sync_on_reconstruction)
+                source, self.mode, self.sync_on_reconstruction)
 
-            if not self.name:
+            if not name:
                 if is_local_source:
                     with ux_utils.print_exception_no_traceback():
                         raise exceptions.StorageNameError(
                             'Storage name must be specified if the source is '
                             'local.')
                 else:
+                    assert isinstance(source, str), source
                     # Set name to source bucket name and continue
-                    self.name = urllib.parse.urlsplit(source).netloc
-                    return
+                    name = urllib.parse.urlsplit(source).netloc
+                    return name, source
             else:
                 if is_local_source:
                     # If name is specified and source is local, upload to bucket
-                    return
+                    return name, source
                 else:
                     # Both name and source should not be specified if the source
                     # is a URI. Name will be inferred from the URI.
@@ -621,6 +632,7 @@ class Storage(object):
             logger.info(f'Storage type {store_type} already exists.')
             return self.stores[store_type]
 
+        store_cls: Type[AbstractStore]
         if store_type == StoreType.S3:
             store_cls = S3Store
         elif store_type == StoreType.GCS:
@@ -630,6 +642,7 @@ class Storage(object):
                 raise exceptions.StorageSpecError(
                     f'{store_type} not supported as a Store.')
 
+        assert self.name is not None, self.name
         # Initialize store object and get/create bucket
         try:
             store = store_cls(name=self.name, source=self.source)
@@ -693,7 +706,8 @@ class Storage(object):
                 store.delete()
                 # Check remaining stores - if none is sky managed, remove
                 # the storage from global_user_state.
-                delete = all(s.is_sky_managed is False for s in self.stores)
+                delete = all(
+                    s.is_sky_managed is False for s in self.stores.values())
                 if delete:
                     global_user_state.remove_storage(self.name)
                 else:
@@ -720,11 +734,19 @@ class Storage(object):
 
     def _sync_store(self, store: AbstractStore):
         """Runs the upload routine for the store and handles failures"""
-        try:
-            if self.source is not None and os.path.isdir(
-                    os.path.join(self.source, '.git')):
+
+        def warn_for_git_dir(source: str):
+            if os.path.isdir(os.path.join(source, '.git')):
                 logger.warning(f'\'.git\' directory under \'{self.source}\' '
                                'is excluded during sync.')
+
+        try:
+            if self.source is not None:
+                if isinstance(self.source, str):
+                    warn_for_git_dir(self.source)
+                else:
+                    for source in self.source:
+                        warn_for_git_dir(source)
             store.upload()
         except exceptions.StorageUploadError:
             logger.error(f'Could not upload {self.source} to store '
@@ -739,7 +761,7 @@ class Storage(object):
             global_user_state.set_storage_status(self.name, StorageStatus.READY)
 
     @classmethod
-    def from_yaml_config(cls, config: Dict[str, str]) -> 'Storage':
+    def from_yaml_config(cls, config: Dict[str, Any]) -> 'Storage':
         backend_utils.validate_schema(config, schemas.get_storage_schema(),
                                       'Invalid storage YAML: ')
 
@@ -774,15 +796,15 @@ class Storage(object):
     def to_yaml_config(self) -> Dict[str, str]:
         config = dict()
 
-        def add_if_not_none(key, value):
+        def add_if_not_none(key: str, value: Optional[Any]):
             if value is not None:
                 config[key] = value
 
-        name = self.name
-        if (self.source is not None and isinstance(self.source, str) and
-                data_utils.is_cloud_store_url(self.source)):
+        name = None
+        if (self.source is None or not isinstance(self.source, str) or
+                not data_utils.is_cloud_store_url(self.source)):
             # Remove name if source is a cloud store URL
-            name = None
+            name = self.name
         add_if_not_none('name', name)
         add_if_not_none('source', self.source)
 
@@ -809,8 +831,8 @@ class S3Store(AbstractStore):
                  source: str,
                  region: Optional[str] = 'us-east-2',
                  is_sky_managed: Optional[bool] = None):
-        self.client = None
-        self.bucket = None
+        self.client: 'boto3.client.Client'
+        self.bucket: 'StorageHandle'
         super().__init__(name, source, region, is_sky_managed)
 
     def _validate(self):
@@ -936,6 +958,7 @@ class S3Store(AbstractStore):
                 max_concurrent_uploads=_MAX_CONCURRENT_UPLOADS)
 
     def _transfer_to_s3(self) -> None:
+        assert isinstance(self.source, str), self.source
         if self.source.startswith('gs://'):
             data_transfer.gcs_to_s3(self.name, self.name)
 
@@ -1081,8 +1104,8 @@ class GcsStore(AbstractStore):
                  source: str,
                  region: Optional[str] = 'us-central1',
                  is_sky_managed: Optional[bool] = None):
-        self.client = None
-        self.bucket = None
+        self.client: 'storage.Client'
+        self.bucket: StorageHandle
         super().__init__(name, source, region, is_sky_managed)
 
     def _validate(self):
@@ -1174,7 +1197,7 @@ class GcsStore(AbstractStore):
         if len(source_path_list) > 1:
             source_message = f'{len(source_path_list)} paths'
         else:
-            source_message = source_path_list
+            source_message = source_path_list[0]
 
         # If the source_path list contains a directory, then gsutil cp -n
         # copies the dir as is to the root of the bucket. To copy the
@@ -1249,7 +1272,7 @@ class GcsStore(AbstractStore):
                 max_concurrent_uploads=_MAX_CONCURRENT_UPLOADS)
 
     def _transfer_to_gcs(self) -> None:
-        if self.source.startswith('s3://'):
+        if isinstance(self.source, str) and self.source.startswith('s3://'):
             data_transfer.s3_to_gcs(self.name, self.name)
 
     def _get_bucket(self) -> Tuple[StorageHandle, bool]:

--- a/sky/data/storage.py
+++ b/sky/data/storage.py
@@ -5,6 +5,7 @@ import subprocess
 import time
 import typing
 from typing import Any, Dict, Optional, Tuple, Type, Union, List
+from typing_extensions import TypeAlias
 import urllib.parse
 
 import colorama
@@ -28,9 +29,10 @@ if typing.TYPE_CHECKING:
 
 logger = sky_logging.init_logger(__name__)
 
-Path = str
 StorageHandle = Any
 StorageStatus = global_user_state.StorageStatus
+Path: TypeAlias = str
+SourceType: TypeAlias = Union[Path, List[Path]]
 
 # Clouds with object storage implemented in this module. Azure Blob
 # Storage isn't supported yet (even though Azure is).
@@ -130,7 +132,7 @@ class AbstractStore:
         def __init__(self,
                      *,
                      name: str,
-                     source: Union[Path, List[Path], None],
+                     source: Optional[SourceType],
                      region: Optional[str] = None,
                      is_sky_managed: Optional[bool] = None):
             self.name = name
@@ -147,7 +149,7 @@ class AbstractStore:
 
     def __init__(self,
                  name: str,
-                 source: Union[Path, List[Path], None],
+                 source: Optional[SourceType],
                  region: Optional[str] = None,
                  is_sky_managed: Optional[bool] = None):
         """Initialize AbstractStore
@@ -302,7 +304,7 @@ class Storage(object):
             self,
             *,
             storage_name: Optional[str],
-            source: Union[str, List[str], None],
+            source: Optional[SourceType],
             sky_stores: Optional[Dict[StoreType,
                                       AbstractStore.StoreMetadata]] = None):
             assert storage_name is not None or source is not None
@@ -329,7 +331,7 @@ class Storage(object):
 
     def __init__(self,
                  name: Optional[str] = None,
-                 source: Union[Path, List[Path], None] = None,
+                 source: Optional[SourceType] = None,
                  stores: Optional[Dict[StoreType, AbstractStore]] = None,
                  persistent: Optional[bool] = True,
                  mode: StorageMode = StorageMode.MOUNT,
@@ -371,6 +373,8 @@ class Storage(object):
             there. This is set to false when the Storage object is created not
             for direct use, e.g. for sky storage delete.
         """
+        self.name = name
+        self.source = source
         self.persistent = persistent
         self.mode = mode
         assert mode in StorageMode
@@ -382,7 +386,8 @@ class Storage(object):
         self.force_delete = False
 
         # Validate and correct inputs if necessary
-        self.name, self.source = self._validate_storage_spec(name, source)
+        self._validate_storage_spec()
+        assert self.name is not None
 
         # Sky optimizer either adds a storage object instance or selects
         # from existing ones
@@ -443,8 +448,8 @@ class Storage(object):
 
     @staticmethod
     def _validate_source(
-            source: Union[str, List[str]], mode: StorageMode,
-            sync_on_reconstruction: bool) -> Tuple[Union[str, List[str]], bool]:
+            source: SourceType, mode: StorageMode,
+            sync_on_reconstruction: bool) -> Tuple[SourceType, bool]:
         """Validates the source path.
 
         Args:
@@ -539,25 +544,22 @@ class Storage(object):
                         f'Supported paths: local, s3://, gs://. Got: {source}')
         return source, is_local_source
 
-    def _validate_storage_spec(
-        self,
-        name: Optional[str],
-        source: Union[Path, List[Path], None],
-    ) -> Tuple[str, Union[Path, List[Path], None]]:
+
+    def _validate_storage_spec(self) -> None:
         """
         Validates the storage spec and updates local fields if necessary.
         """
-        if source is None:
+        if self.source is None:
             # If the mode is COPY, the source must be specified
             if self.mode == StorageMode.COPY:
                 # Check if a Storage object already exists in global_user_state
                 # (e.g. used as scratch previously). Such storage objects can be
                 # mounted in copy mode even though they have no source in the
                 # yaml spec (the name is the source).
-                handle = global_user_state.get_handle_from_storage_name(name)
+                handle = global_user_state.get_handle_from_storage_name(
+                    self.name)
                 if handle is not None:
-                    assert name is not None
-                    return name, source
+                    return
                 else:
                     with ux_utils.print_exception_no_traceback():
                         raise exceptions.StorageSourceError(
@@ -567,32 +569,31 @@ class Storage(object):
                 # If source is not specified in COPY mode, the intent is to
                 # create a bucket and use it as scratch disk. Name must be
                 # specified to create bucket.
-                if not name:
+                if not self.name:
                     with ux_utils.print_exception_no_traceback():
                         raise exceptions.StorageSpecError(
                             'Storage source or storage name must be specified.')
                 else:
                     # Create bucket and mount
-                    return name, source
-        elif source is not None:
+                    return
+        elif self.source is not None:
             source, is_local_source = Storage._validate_source(
-                source, self.mode, self.sync_on_reconstruction)
-
-            if not name:
+                self.source, self.mode, self.sync_on_reconstruction)
+            if not self.name:
                 if is_local_source:
                     with ux_utils.print_exception_no_traceback():
                         raise exceptions.StorageNameError(
                             'Storage name must be specified if the source is '
                             'local.')
                 else:
-                    assert isinstance(source, str), source
+                    assert isinstance(source, str)
                     # Set name to source bucket name and continue
-                    name = urllib.parse.urlsplit(source).netloc
-                    return name, source
+                    self.name = urllib.parse.urlsplit(source).netloc
+                    return
             else:
                 if is_local_source:
                     # If name is specified and source is local, upload to bucket
-                    return name, source
+                    return
                 else:
                     # Both name and source should not be specified if the source
                     # is a URI. Name will be inferred from the URI.

--- a/sky/data/storage.py
+++ b/sky/data/storage.py
@@ -226,25 +226,14 @@ class AbstractStore:
         """
         raise NotImplementedError
 
-    # def download_remote_dir(self, local_path: str) -> None:
-    #     """Downloads directory from remote bucket to the specified
-    #     local_path
+    def download_remote_dir(self, local_path: str) -> None:
+        """Downloads directory from remote bucket to the specified
+        local_path
 
-    #     Args:
-    #       local_path: Local path on user's device
-    #     """
-    #     assert local_path is not None
-    #     local_path = os.path.expanduser(local_path)
-    #     iterator = self._remote_filepath_iterator()
-    #     for remote_path in iterator:
-    #         remote_path = next(iterator)
-    #         if remote_path[-1] == '/':
-    #             continue
-    #         path = os.path.join(local_path, remote_path)
-    #         if not os.path.exists(os.path.dirname(path)):
-    #             os.makedirs(os.path.dirname(path))
-    #         logger.info(f'Downloading {remote_path} to {path}')
-    #         self._download_file(remote_path, path)
+        Args:
+          local_path: Local path on user's device
+        """
+        raise NotImplementedError
 
     def _download_file(self, remote_path: str, local_path: str) -> None:
         """Downloads file from remote to local on Store

--- a/sky/data/storage.py
+++ b/sky/data/storage.py
@@ -544,7 +544,6 @@ class Storage(object):
                         f'Supported paths: local, s3://, gs://. Got: {source}')
         return source, is_local_source
 
-
     def _validate_storage_spec(self) -> None:
         """
         Validates the storage spec and updates local fields if necessary.

--- a/sky/data/storage.py
+++ b/sky/data/storage.py
@@ -391,10 +391,11 @@ class Storage(object):
         # Logic to rebuild Storage if it is in global user state
         handle = global_user_state.get_handle_from_storage_name(self.name)
         if handle is not None:
+            self.handle = handle 
             # Reconstruct the Storage object from the global_user_state
             logger.debug('Detected existing storage object, '
                          f'loading Storage: {self.name}')
-            for s_type, s_metadata in handle.sky_stores.items():
+            for s_type, s_metadata in self.handle.sky_stores.items():
                 # When initializing from global_user_state, we override the
                 # source from the YAML
                 if s_type == StoreType.S3:
@@ -427,7 +428,7 @@ class Storage(object):
                 for t, s in self.stores.items()
                 if s.is_sky_managed
             }
-            handle = self.StorageMetadata(storage_name=self.name,
+            self.handle = self.StorageMetadata(storage_name=self.name,
                                           source=self.source,
                                           sky_stores=sky_managed_stores)
 
@@ -439,8 +440,6 @@ class Storage(object):
                         self.add_store(StoreType.S3)
                     elif self.source.startswith('gs://'):
                         self.add_store(StoreType.GCS)
-        assert handle is not None, 'handle should not be None'
-        self.handle = handle
 
     @staticmethod
     def _validate_source(

--- a/sky/data/storage.py
+++ b/sky/data/storage.py
@@ -391,7 +391,7 @@ class Storage(object):
         # Logic to rebuild Storage if it is in global user state
         handle = global_user_state.get_handle_from_storage_name(self.name)
         if handle is not None:
-            self.handle = handle 
+            self.handle = handle
             # Reconstruct the Storage object from the global_user_state
             logger.debug('Detected existing storage object, '
                          f'loading Storage: {self.name}')
@@ -429,8 +429,8 @@ class Storage(object):
                 if s.is_sky_managed
             }
             self.handle = self.StorageMetadata(storage_name=self.name,
-                                          source=self.source,
-                                          sky_stores=sky_managed_stores)
+                                               source=self.source,
+                                               sky_stores=sky_managed_stores)
 
             if self.source is not None:
                 # If source is a pre-existing bucket, connect to the bucket

--- a/sky/global_user_state.py
+++ b/sky/global_user_state.py
@@ -374,11 +374,15 @@ def set_storage_handle(storage_name: str, handle: 'Storage.StorageMetadata'):
         raise ValueError(f'Storage{storage_name} not found.')
 
 
-def get_handle_from_storage_name(storage_name: str):
-    assert storage_name is not None, 'storage_name cannot be None'
+def get_handle_from_storage_name(
+        storage_name: Optional[str]) -> Optional['Storage.StorageMetadata']:
+    if storage_name is None:
+        return None
     rows = _DB.cursor.execute('SELECT handle FROM storage WHERE name=(?)',
                               (storage_name,))
     for (handle,) in rows:
+        if handle is None:
+            return None
         return pickle.loads(handle)
 
 

--- a/tests/mypy_files.txt
+++ b/tests/mypy_files.txt
@@ -1,0 +1,1 @@
+sky/data/storage.py

